scrapy/utils/datatypes.py

```diff
@@ -20,6 +20,13 @@
     Mapping = collections.abc.Mapping
 
 
+# See https://twistedmatrix.com/documents/current/api/twisted.internet.address.html
+IPv4Address = collections.namedtuple('IPv4Address', ['type', 'host', 'port'])
```

Comment:
one more option is to use https://docs.python.org/3/library/ipaddress.html#ip-addresses; it is new in 3.3, but there can be backport available

Reply:
I think that's fine not to return type and port, as they should be available from url (?). In stdlib ip address data structure doesn't include port and type - and it makes sense to me, as they are not a part of ip address

Reply:
Here I am re-inventing the wheel :-)

I have a few conflicting ideas about this. On the one hand, why not expose information that we have available, like type and port? The overhead of the additional classes is minimal, since they are `namedtuples`. On the other hand, I understand that if the idea is to be Twisted-independent, an eventual future implementation might not expose that, and we'd need to do backward incompatible changes. But then again, how far are we from moving away from Twisted? I wonder if that is perhaps too much early optimization :stuck_out_tongue: 

In addition, I think adding a 3rd party library backport for py2 is in fact a bit of an overkill. What about returning an `ipaddress.IPv4Address` in py3 and a plain string in py2?

Reply:
one more option: wait for 1.8 (we aim for August), drop Python 2 support in master, merge afterwards

Reply:
Yes, that sounds reasonable. Let me change the title and add this to the 2.0 milestone. Thanks for the feedback :-)

Reply:
Done. py2 tests are failing, which is expected.